pkg/http/auth_middleware.go

```diff
@@ -0,0 +1,352 @@
+package http
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"os"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+	"unicode/utf8"
+)
+
+// Signature and API related constants.
+const (
+	unsignedPayload   = "UNSIGNED-PAYLOAD"
+	signV4Algorithm   = "AWS4-HMAC-SHA256"
+	iso8601DateFormat = "20060102T150405Z"
+	yyyymmdd          = "20060102"
+)
+
+var reservedObjectNames = regexp.MustCompile("^[a-zA-Z0-9-_.~/]+$")
+
+type credential struct {
+	AccessKeyID string
+	Time        string
+	Location    string
+	Service     string
+	Request     string
+}
+
+func parseCredential(str string) (credential, error) {
+	parts := strings.Split(str, "/")
+	if len(parts) != 5 {
+		return credential{}, fmt.Errorf("invalid credential format")
+	}
+
+	return credential{
+		AccessKeyID: parts[0],
+		Time:        parts[1],
+		Location:    parts[2],
+		Service:     parts[3],
+		Request:     parts[4],
+	}, nil
+}
+
+func authMiddleware(handler http.Handler) http.Handler {
```

Comment:
How about this instead so the passing of the AWS credentials isn't hidden in the func implementation but rather is part of its signature:
```suggestion
func authMiddleware(awsKeyID, awsSecretKey string, handler http.Handler) http.Handler {
```